File: LeetCode/Graph/BFS/JS/Leetcode654-maximum-binary-tree.js
var constructMaximumBinaryTree = function(nums) {
    let n = nums.length;
    if(!n) return null;
    let dfs = function(num,l,r) {
        if(l == r) {
            return new TreeNode(nums[l]);
        }
        if(l > r) return null;
        let [max_v,idx] = [0,l];
        for(let i = l;i <= r;i++) {
            if(max_v < nums[i]) {
                max_v = nums[i];
                idx = i;
            }
        }
        let node = new TreeNode(max_v);
        node.left = dfs(nums,l,idx-1);
        node.right = dfs(nums,idx + 1,r);
        return node;
    }
    return dfs(nums,0,n - 1);
};

File: LeetCode/Graph/BFS/Golang/Leetcode654-maximum-binary-tree.go
package sbgo

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func constructMaximumBinaryTree(nums []int) *TreeNode {
	n := len(nums)
	if n == 0 {
		return nil
	}
	var dfs func(nums []int, l, r int) *TreeNode
	dfs = func(nums []int, l, r int) *TreeNode {
		if l > r {
			return nil
		}
		if l == r {
			return &TreeNode{nums[l], nil, nil}
		}
		max_v, idx := 0, l
		for i := l; i <= r; i++ {
			if max_v < nums[i] {
				max_v = nums[i]
				idx = i
			}
		}
		node := TreeNode{max_v, nil, nil}
		node.Left = dfs(nums, l, idx-1)
		node.Right = dfs(nums, idx+1, r)
		return &node
	}
	return dfs(nums, 0, n-1)
}
